emoji/src/main/java/com/gluonhq/emoji/DownloadableEmojiSpriteLoader.java

```diff
@@ -0,0 +1,104 @@
+package com.gluonhq.emoji;
+
+import javafx.scene.image.Image;
+
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Logger;
+
+public class DownloadableEmojiSpriteLoader implements EmojiSpriteLoader {
+
+    private static final Logger LOG = Logger.getLogger(DownloadableEmojiSpriteLoader.class.getName());
+
+    private static final String COMMIT_NUMBER = "063f328d7951cb2e2a6678b06dcbdf8dd599fad7"; // tag 15.0.1
+    private static final String EMOJI_PNG_URL = "https://github.com/iamcal/emoji-data/blob/" + COMMIT_NUMBER + "/sheets-clean/sheet_apple_%s_clean.png?raw=true";
+
+    private static final String LOCAL_PATH = System.getProperty("user.home") + "/.gluon/emoji/" + COMMIT_NUMBER;
+    private static final int[] EMOJI_SIZES = new int[] { 20, 32, 64 };
+    private boolean initialized;
+
+    @Override
+    public boolean isInitialized() {
+        if (!initialized) {
+            for (int size : EMOJI_SIZES) {
+                String fileName = "sheet_apple_" + size + ".png";
+                if (!Files.exists(Paths.get(LOCAL_PATH, fileName))) {
+                    return false;
+                }
+            }
+        }
+        return initialized = true;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> initialize() {
+        return CompletableFuture.supplyAsync(() -> {
```

Comment:
When using CompletableFuture, it's best to run the runnables in a custom thread pool instead of relying on the JDK's common pool when the runnable does blocking I/O operations (i.e. downloading files).

Reply:
Does using a `Executors.newSingleThreadExecutor()` makes sense here?

```
public CompletableFuture<Boolean> initialize() {
    return CompletableFuture.supplyAsync(() -> {
        ...
    }, Executors.newSingleThreadExecutor());
}```

Reply:
That should indeed be sufficient. The reason to run the runnable in its own executor, is that it won't block other runnables in the application that might use the same common pool while the download is in progress.